File: coolero/repositories/composite_repo.py
#  Coolero - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging
from statistics import mean
from typing import List, Tuple, Set

import matplotlib
import numpy

from coolero.models.device import Device, DeviceType
from coolero.models.device_info import DeviceInfo
from coolero.models.status import Status, TempStatus
from coolero.repositories.cpu_repo import CPU_TEMP
from coolero.repositories.devices_repository import DevicesRepository
from coolero.repositories.gpu_repo import GPU_TEMP

_LOG = logging.getLogger(__name__)
_AVG_ALL: str = 'Average All'
_LIQUID_TEMP_NAMES: Tuple[str, str] = ('Liquid', 'Water')


class CompositeRepo(DevicesRepository):
    """This class if for combining different temp sources from devices for lots of possibilities"""
    _composite_device_status: Device | None

    def __init__(self, devices: List[Device]) -> None:
        self._devices = devices
        super().__init__()
        _LOG.info('Initialized with status: %s', self._composite_device_status)

    def _initialize_devices(self) -> None:
        if len(self._devices) < 2:
            return
        if all_temps := self._collect_all_temps():
            avg_all_temps = self._get_avg_all_temps(all_temps)
            delta_cpu_temps = self._get_delta_cpu_liquid_temps(all_temps)
            delta_gpu_temps = self._get_delta_gpu_liquid_temps(all_temps)
            composite_temps = avg_all_temps + delta_cpu_temps + delta_gpu_temps
            colors = self._get_device_colors(len(composite_temps))
            self._composite_device_status = Device(
                _name='Composite',
                _type_id=(DeviceType.COMPOSITE, 1),
                _status_current=Status(temps=composite_temps),
                _colors={
                    temp_status.name: colors[index]
                    for index, temp_status in enumerate(composite_temps)
                },
                _info=DeviceInfo(
                    temp_min=0,
                    temp_max=100,
                    temp_ext_available=True,
                    profile_max_length=21
                )
            )

    @property
    def statuses(self) -> List[Device]:
        return [self._composite_device_status]

    def update_statuses(self) -> None:
        if self._composite_device_status is None:
            return
        if all_temps := self._collect_all_temps():
            avg_all_temps = self._get_avg_all_temps(all_temps)
            delta_cpu_temps = self._get_delta_cpu_liquid_temps(all_temps)
            delta_gpu_temps = self._get_delta_gpu_liquid_temps(all_temps)
            composite_temps = avg_all_temps + delta_cpu_temps + delta_gpu_temps
            self._composite_device_status.status = Status(temps=composite_temps)
            _LOG.debug('Composite device: %s status was updated with status: %s',
                       self._composite_device_status.name,
                       self._composite_device_status.status)

    def shutdown(self) -> None:
        self._composite_device_status = None
        _LOG.debug("Composite Repo shutdown")

    def _collect_all_temps(self) -> Set[Tuple[str, float, int]]:
        all_temps = {
            (temp_status.external_name, temp_status.temp, device.type_id)
            for device in self._devices
            if device.type != DeviceType.COMPOSITE
            for temp_status in device.status.temps
        }
        return all_temps if len(all_temps) > 1 else set()

    @staticmethod
    def _get_avg_all_temps(all_temps: Set[Tuple[str, float, int]]) -> List[TempStatus]:
        _, temps, _ = zip(*all_temps)
        return [TempStatus(_AVG_ALL, mean(temps), _AVG_ALL, _AVG_ALL)]

    @staticmethod
    def _get_delta_cpu_liquid_temps(all_temps: Set[Tuple[str, float, int]]) -> List[TempStatus]:
        deltas: List[TempStatus] = []
        if cpu_temp := next((temp for name, temp, _ in all_temps if name == CPU_TEMP), None):
            liquid_temps = filter(
                lambda name_temp: _LIQUID_TEMP_NAMES[0] in name_temp[0] or _LIQUID_TEMP_NAMES[1] in name_temp[0],
                all_temps
            )
            for name, temp, _ in liquid_temps:
                delta_temp_name = f'Δ CPU {name}'
                deltas.append(
                    TempStatus(delta_temp_name, abs(cpu_temp - temp), delta_temp_name, delta_temp_name)
                )
        return deltas

    @staticmethod
    def _get_delta_gpu_liquid_temps(all_temps: Set[Tuple[str, float, int]]) -> List[TempStatus]:
        deltas: List[TempStatus] = []
        if gpu_temps := [(temp, dev_id) for name, temp, dev_id in all_temps if name == GPU_TEMP]:
            liquid_temps = filter(
                lambda name_temp_id: _LIQUID_TEMP_NAMES[0] in name_temp_id[0]
                                     or _LIQUID_TEMP_NAMES[1] in name_temp_id[0],
                all_temps
            )
            number_gpus = len(gpu_temps)
            for name, temp, device_id in liquid_temps:
                for gpu_temp, gpu_id in gpu_temps:
                    delta_temp_name = f'Δ GPU {name}' if number_gpus == 1 else f'Δ GPU#{gpu_id} {name}'
                    deltas.append(
                        TempStatus(delta_temp_name, abs(gpu_temp - temp), delta_temp_name, delta_temp_name)
                    )
        return deltas

    @staticmethod
    def _get_device_colors(number_of_colors: int) -> List[str]:
        colors_selectors = numpy.linspace(0.5, 0.9, number_of_colors)
        color_map = matplotlib.cm.get_cmap('copper')(colors_selectors)
        return [matplotlib.cm.colors.to_hex(color) for color in color_map]

File: coolercontrold/src/repositories/composite_repo.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2023  Guy Boldon
 * |
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * |
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * |
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::sync::Arc;

use anyhow::{anyhow, Result};
use async_trait::async_trait;
use log::{debug, info, trace};
use tokio::sync::RwLock;
use tokio::time::Instant;

use crate::config::Config;
use crate::device::{Device, DeviceInfo, DeviceType, Status, TempStatus, UID};
use crate::repositories::repository::{DeviceList, DeviceLock, Repository};
use crate::setting::{LcdSettings, LightingSettings, TempSource};

const AVG_ALL: &str = "Average All";
const MAX_ALL: &str = "Max All";

type AllTemps = Vec<(String, f64, u8)>;

/// A Repository for Composite Temperatures of other repositories
//#[deprecated(since="0.18.0", note="To be removed after deployment of custom sensors")]
pub struct CompositeRepo {
    config: Arc<Config>,
    composite_device: Option<DeviceLock>,
    other_devices: DeviceList,
    liquid_temp_names: Vec<String>,
}

impl CompositeRepo {
    pub fn new(config: Arc<Config>, devices_for_composite: DeviceList) -> Self {
        Self {
            config,
            composite_device: None,
            other_devices: devices_for_composite,
            liquid_temp_names: vec!["Liquid".to_string(), "Water".to_string()],
        }
    }

    async fn collect_all_temps(&self) -> AllTemps {
        let mut all_temps = Vec::new();
        for device in self.other_devices.iter() {
            let type_index = device.read().await.type_index;
            let current_status = &device.read().await.status_current();
            if let Some(status) = current_status {
                for temp_status in &status.temps {
                    all_temps.push((temp_status.external_name.clone(), temp_status.temp, type_index));
                }
            }
        }
        all_temps
    }

    fn get_avg_all_temps(&self, all_temps: &AllTemps) -> Vec<TempStatus> {
        let total_all_temps: f64 = all_temps.iter()
            .map(|(_, temp, _)| temp)
            .sum();
        let average = (total_all_temps / all_temps.len() as f64 * 100.0).round() / 100.0;
        let temp_status = TempStatus {
            name: AVG_ALL.to_string(),
            temp: average,
            frontend_name: AVG_ALL.to_string(),
            external_name: AVG_ALL.to_string(),
        };
        vec![temp_status]
    }

    fn get_max_all_temps(&self, all_temps: &AllTemps) -> Vec<TempStatus> {
        let max_all_temps: f64 = all_temps.iter()
            .map(|(_, temp, _)| temp)
            .max_by(|a, b| a.total_cmp(b))
            .map(|temp| temp.clone())
            .unwrap_or_default();
        let temp_status = TempStatus {
            name: MAX_ALL.to_string(),
            temp: max_all_temps,
            frontend_name: MAX_ALL.to_string(),
            external_name: MAX_ALL.to_string(),
        };
        vec![temp_status]
    }

    fn get_delta_cpu_liquid_temps(&self, all_temps: &AllTemps) -> Vec<TempStatus> {
        let mut deltas = Vec::new();
        let cpu_temps = all_temps.iter()
            .filter(|(external_name, _, _)| external_name.contains("CPU"))
            .collect::<Vec<&(String, f64, u8)>>();
        all_temps.iter()
            .filter(|(name, _, _)|
                self.liquid_temp_names.iter().any(|liquid_temp_name| name.contains(liquid_temp_name))
            ).for_each(|(liquid_name, liquid_temp, _)| {
            for (cpu_name, cpu_temp, _) in &cpu_temps {
                let delta_temp_name = format!("Δ {} {}", cpu_name, liquid_name);
                deltas.push(
                    TempStatus {
                        name: delta_temp_name.clone(),
                        temp: ((cpu_temp - liquid_temp).abs() * 100.0).round() / 100.0,
                        frontend_name: delta_temp_name.clone(),
                        external_name: delta_temp_name,
                    }
                );
            }
        });
        deltas
    }

    fn get_delta_gpu_liquid_temps(&self, all_temps: &AllTemps) -> Vec<TempStatus> {
        let mut deltas = Vec::new();
        let gpu_temps = all_temps.iter()
            .filter(|(external_name, _, _)| external_name.contains("GPU"))
            .collect::<Vec<&(String, f64, u8)>>();
        all_temps.iter()
            .filter(|(name, _, _)|
                self.liquid_temp_names.iter().any(|liquid_temp_name| name.contains(liquid_temp_name))
            ).for_each(|(liquid_name, liquid_temp, _)| {
            for (gpu_name, gpu_temp, _) in &gpu_temps {
                let delta_temp_name = format!("Δ {} {}", gpu_name, liquid_name);
                deltas.push(
                    TempStatus {
                        name: delta_temp_name.clone(),
                        temp: ((gpu_temp - liquid_temp).abs() * 100.0).round() / 100.0,
                        frontend_name: delta_temp_name.clone(),
                        external_name: delta_temp_name,
                    }
                );
            }
        });
        deltas
    }

    fn get_max_cpu_gpu_temps(&self, all_temps: &AllTemps) -> Vec<TempStatus> {
        let mut cpu_gpu_maximums = Vec::new();
        let base_cpu_temps = all_temps.iter()
            .filter(|(external_name, _, _)|
                external_name.contains("CPU") &&
                    (external_name.contains("Package") || external_name.contains("Tctl"))
            ).collect::<Vec<&(String, f64, u8)>>();
        all_temps.iter()
            .filter(|(external_name, _, _)| external_name.contains("GPU"))
            .for_each(|(gpu_name, gpu_temp, _)| {
                for (cpu_name, cpu_temp, _) in &base_cpu_temps {
                    let max_temp_name = format!("Max {} {}", cpu_name, gpu_name);
                    cpu_gpu_maximums.push(
                        TempStatus {
                            name: max_temp_name.clone(),
                            temp: cpu_temp.clone().max(gpu_temp.clone()),
                            frontend_name: max_temp_name.clone(),
                            external_name: max_temp_name,
                        }
                    );
                }
            });
        cpu_gpu_maximums
    }
}

#[async_trait]
impl Repository for CompositeRepo {
    fn device_type(&self) -> DeviceType {
        DeviceType::Composite
    }

    async fn initialize_devices(&mut self) -> Result<()> {
        debug!("Starting Device Initialization");
        let start_initialization = Instant::now();
        let composite_device = Device::new(
            "Composite".to_string(),
            DeviceType::Composite,
            1,
            None,
            Some(DeviceInfo {
                temp_min: 0,
                temp_max: 100,
                temp_ext_available: true,
                profile_max_length: 21,
                ..Default::default()
            }),
            None,
        );
        let cc_device_setting = self.config.get_cc_settings_for_device(
            &composite_device.uid
        ).await?;
        if cc_device_setting.is_some() && cc_device_setting.unwrap().disable {
            info!("Skipping disabled composite device with UID: {}", composite_device.uid);
        } else if self.other_devices.len() > 1 {
            self.composite_device = Some(Arc::new(RwLock::new(composite_device)));
        }
        self.update_statuses().await?;
        if self.composite_device.is_some() {
            let recent_status = self.composite_device
            .as_ref()
            .unwrap()
            .read()
            .await
            .status_current()
            .unwrap();
            self.composite_device
            .as_ref()
            .unwrap()
            .write()
            .await
            .initialize_status_history_with(recent_status);
        }
        if log::max_level() == log::LevelFilter::Debug {
            if let Some(composite_device) = self.composite_device.as_ref() {
                info!("Initialized Composite Device: {:?}", composite_device.read().await);
            } else {
                info!("Initialized Composite Device: None");
            }
        }
        trace!(
            "Time taken to initialize COMPOSITE device: {:?}", start_initialization.elapsed()
        );
        debug!("COMPOSITE Repository initialized");
        Ok(())
    }

    async fn devices(&self) -> DeviceList {
        if let Some(device) = self.composite_device.as_ref() {
            vec![device.clone()]
        } else {
            Vec::new()
        }
    }

    /// For composite repos, there is no need to preload as other device statuses
    /// have already been updated.
    async fn preload_statuses(self: Arc<Self>) {}

    async fn update_statuses(&self) -> Result<()> {
        if let Some(composite_device) = self.composite_device.as_ref() {
            let start_update = Instant::now();
            let all_temps = self.collect_all_temps().await;
            if all_temps.len() > 1 {
                let mut composite_temps = Vec::new();
                composite_temps.append(&mut self.get_avg_all_temps(&all_temps));
                composite_temps.append(&mut self.get_max_all_temps(&all_temps));
                composite_temps.append(&mut self.get_delta_cpu_liquid_temps(&all_temps));
                composite_temps.append(&mut self.get_delta_gpu_liquid_temps(&all_temps));
                composite_temps.append(&mut self.get_max_cpu_gpu_temps(&all_temps));
                composite_device.write().await.set_status(
                    Status {
                        temps: composite_temps,
                        ..Default::default()
                    }
                )
            }
            trace!(
                "STATUS SNAPSHOT Time taken for COMPOSITE device: {:?}",
                start_update.elapsed()
            );
        }
        Ok(())
    }

    async fn shutdown(&self) -> Result<()> {
        info!("COMPOSITE Repository shutdown");
        Ok(())
    }

    async fn apply_setting_reset(&self, _device_uid: &UID, _channel_name: &str) -> Result<()> {
        Ok(())
    }
    async fn apply_setting_speed_fixed(&self, _device_uid: &UID, _channel_name: &str, _speed_fixed: u8) -> Result<()> {
        Err(anyhow!("Applying settings Speed Fixed is not supported for COMPOSITE devices"))
    }
    async fn apply_setting_speed_profile(&self, _device_uid: &UID, _channel_name: &str, _temp_source: &TempSource, _speed_profile: &Vec<(f64, u8)>) -> Result<()> {
        Err(anyhow!("Applying settings Speed Profile is not supported for COMPOSITE devices"))
    }
    async fn apply_setting_lighting(&self, _device_uid: &UID, _channel_name: &str, _lighting: &LightingSettings) -> Result<()> {
        Err(anyhow!("Applying settings Lighting is not supported for COMPOSITE devices"))
    }
    async fn apply_setting_lcd(&self, _device_uid: &UID, _channel_name: &str, _lcd: &LcdSettings) -> Result<()> {
        Err(anyhow!("Applying settings LCD is not supported for COMPOSITE devices"))
    }
    async fn apply_setting_pwm_mode(&self, _device_uid: &UID, _channel_name: &str, _pwm_mode: u8) -> Result<()> {
        Err(anyhow!("Applying settings pwm_mode is not supported for COMPOSITE devices"))
    }
}
